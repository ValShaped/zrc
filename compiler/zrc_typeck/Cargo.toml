[package]
name = "zrc_typeck"
version = "0.1.0"
edition = "2024"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
derive_more = { version = "2.0.1", features = ["display"] }
indexmap = "2.2.6"
zrc_diagnostics = { version = "0.1.0", path = "../zrc_diagnostics" }
zrc_parser = { version = "0.1.0", path = "../zrc_parser" }
zrc_utils = { version = "0.1.0", path = "../zrc_utils" }
