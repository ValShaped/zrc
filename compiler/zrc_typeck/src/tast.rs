//! Type-checked AST for the Zirco programming language
//!
//! This is similar to the [standard AST](zrc_parser::ast), but with type
//! information attached to it.

pub mod expr;
pub mod stmt;
pub mod ty;
