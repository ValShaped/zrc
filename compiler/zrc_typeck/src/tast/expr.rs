//! Expression representation for the Zirco [TAST](super)

use std::fmt::Display;

/// An [expression kind](TypedExprKind) with its yielded [result type](super::ty::Type) attached to it.
#[derive(PartialEq, Debug, Clone)]
#[allow(clippy::module_name_repetitions)]
pub struct TypedExpr(pub super::ty::Type, pub TypedExprKind);

/// The kind of a [`TypedExpr`]
#[derive(PartialEq, Debug, Clone)]
pub enum TypedExprKind {
    /// `a, b`
    Comma(Box<TypedExpr>, Box<TypedExpr>),

    /// `a = b`
    Assignment(Box<TypedExpr>, Box<TypedExpr>),

    /// `!x`
    UnaryNot(Box<TypedExpr>),
    /// `~x`
    UnaryBitwiseNot(Box<TypedExpr>),
    /// `-x`
    UnaryMinus(Box<TypedExpr>),
    /// `&x`
    UnaryAddressOf(Box<TypedExpr>),
    /// `*x`
    UnaryDereference(Box<TypedExpr>),

    /// `a[b]`
    Index(Box<TypedExpr>, Box<TypedExpr>),
    /// `a.b`
    Dot(Box<TypedExpr>, String),
    /// `a->b`
    Arrow(Box<TypedExpr>, String),
    /// `a(b, c, d, ...)`
    Call(Box<TypedExpr>, Vec<TypedExpr>),

    /// `a ? b : c`
    Ternary(Box<TypedExpr>, Box<TypedExpr>, Box<TypedExpr>),

    /// `a && b`
    LogicalAnd(Box<TypedExpr>, Box<TypedExpr>),
    /// `a || b`
    LogicalOr(Box<TypedExpr>, Box<TypedExpr>),

    /// `a == b`
    Equals(Box<TypedExpr>, Box<TypedExpr>),
    /// `a != b`
    NotEquals(Box<TypedExpr>, Box<TypedExpr>),

    /// `a & b`
    BitwiseAnd(Box<TypedExpr>, Box<TypedExpr>),
    /// `a | b`
    BitwiseOr(Box<TypedExpr>, Box<TypedExpr>),
    /// `a ^ b`
    BitwiseXor(Box<TypedExpr>, Box<TypedExpr>),

    /// `a > b`
    GreaterThan(Box<TypedExpr>, Box<TypedExpr>),
    /// `a >= b`
    GreaterThanOrEqualTo(Box<TypedExpr>, Box<TypedExpr>),
    /// `a < b`
    LessThan(Box<TypedExpr>, Box<TypedExpr>),
    /// `a <= b`
    LessThanOrEqualTo(Box<TypedExpr>, Box<TypedExpr>),

    /// `a >> b`
    BitwiseRightShift(Box<TypedExpr>, Box<TypedExpr>),
    /// `a << b`
    BitwiseLeftShift(Box<TypedExpr>, Box<TypedExpr>),

    /// `a + b`
    Addition(Box<TypedExpr>, Box<TypedExpr>),
    /// `a - b`
    Subtraction(Box<TypedExpr>, Box<TypedExpr>),

    /// `a * b`
    Multiplication(Box<TypedExpr>, Box<TypedExpr>),
    /// `a / b`
    Division(Box<TypedExpr>, Box<TypedExpr>),
    /// `a % b`
    Modulo(Box<TypedExpr>, Box<TypedExpr>),

    /// `x as T`
    Cast(Box<TypedExpr>, super::ty::Type),

    /// Any numeric literal.
    NumberLiteral(String),
    /// Any string literal.
    StringLiteral(String),
    /// Any identifier.
    Identifier(String),
    /// Any boolean literal.
    BooleanLiteral(bool),
}

impl Display for TypedExprKind {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::NumberLiteral(n) => write!(f, "{n}"),
            Self::StringLiteral(s) => write!(f, "{s}"),
            Self::Identifier(i) => write!(f, "{i}"),
            Self::BooleanLiteral(b) => write!(f, "{b}"),
            Self::Assignment(l, r) => write!(f, "{l} = {r}"),
            Self::Addition(l, r) => write!(f, "{l} + {r}"),
            Self::Subtraction(l, r) => write!(f, "{l} - {r}"),
            Self::Multiplication(l, r) => write!(f, "{l} * {r}"),
            Self::Division(l, r) => write!(f, "{l} / {r}"),
            Self::Modulo(l, r) => write!(f, "{l} % {r}"),
            Self::BitwiseAnd(l, r) => write!(f, "{l} & {r}"),
            Self::BitwiseOr(l, r) => write!(f, "{l} | {r}"),
            Self::BitwiseXor(l, r) => write!(f, "{l} ^ {r}"),
            Self::BitwiseLeftShift(l, r) => write!(f, "{l} << {r}"),
            Self::BitwiseRightShift(l, r) => write!(f, "{l} >> {r}"),
            Self::GreaterThan(l, r) => write!(f, "{l} > {r}"),
            Self::GreaterThanOrEqualTo(l, r) => write!(f, "{l} >= {r}"),
            Self::LessThan(l, r) => write!(f, "{l} < {r}"),
            Self::LessThanOrEqualTo(l, r) => write!(f, "{l} <= {r}"),
            Self::Equals(l, r) => write!(f, "{l} == {r}"),
            Self::NotEquals(l, r) => write!(f, "{l} != {r}"),
            Self::LogicalAnd(l, r) => write!(f, "{l} && {r}"),
            Self::LogicalOr(l, r) => write!(f, "{l} || {r}"),
            Self::Comma(l, r) => write!(f, "{l}, {r}"),
            Self::UnaryNot(e) => write!(f, "!{e}"),
            Self::Cast(x, t) => write!(f, "{x} as {t}"),
            Self::UnaryBitwiseNot(e) => write!(f, "~{e}"),
            Self::UnaryMinus(e) => write!(f, "-{e}"),
            Self::UnaryAddressOf(e) => write!(f, "&{e}"),
            Self::UnaryDereference(e) => write!(f, "*{e}"),
            Self::Arrow(l, r) => write!(f, "{l}->{r}"),
            Self::Ternary(a, b, c) => write!(f, "{a} ? {b} : {c}"),
            Self::Index(a, b) => write!(f, "{a}[{b}]"),
            Self::Dot(a, b) => write!(f, "{a}.{b}"),
            Self::Call(a, b) => write!(
                f,
                "{a}({})",
                b.iter()
                    .map(ToString::to_string)
                    .collect::<Vec<String>>()
                    .join(", ")
            ),
        }
    }
}

impl Display for TypedExpr {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "(({}) as ({}))", self.1, self.0)
    }
}
