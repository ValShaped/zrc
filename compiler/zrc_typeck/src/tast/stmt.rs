//! Statement representation for the Zirco [TAST](super)

use std::fmt::Display;

use indexmap::IndexMap;

use super::{expr::TypedExpr, ty::Type};

/// A declaration created with `let`.
#[derive(Debug, Clone, PartialEq)]
pub struct LetDeclaration {
    /// The name of the identifier.
    pub name: String,
    /// The type of the new symbol. If set to [`None`], the type will be
    /// inferred.
    pub ty: Type, // types are definite after inference
    /// The value to associate with the new symbol.
    pub value: Option<TypedExpr>,
}

impl Display for LetDeclaration {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match &self.value {
            Some(v) => write!(f, "{}: {} = {}", self.name, self.ty, v),
            None => write!(f, "{}: {}", self.name, self.ty),
        }
    }
}

/// The enum representing all of the different kinds of statements in Zirco
/// after type checking
#[derive(Debug, Clone, PartialEq)]
#[allow(clippy::module_name_repetitions)]
pub enum TypedStmt {
    // all of the Box<Stmt>s for "possibly blocks" have been desugared into vec[single stmt] here
    // (basically if (x) y has become if (x) {y})
    /// `if (x) y` or `if (x) y else z`
    IfStmt(TypedExpr, Vec<TypedStmt>, Option<Vec<TypedStmt>>),
    /// `while (x) y`
    WhileStmt(TypedExpr, Vec<TypedStmt>),
    /// `for (init; cond; post) body`
    ForStmt {
        /// Runs once before the loop starts.
        init: Option<Box<Vec<LetDeclaration>>>,
        /// Runs before each iteration of the loop. If this evaluates to
        /// `false`, the loop will end. If this is [`None`], the loop
        /// will run forever.
        cond: Option<TypedExpr>,
        /// Runs after each iteration of the loop.
        post: Option<TypedExpr>,
        /// The body of the loop.
        body: Vec<TypedStmt>,
    },
    /// `{ ... }`
    BlockStmt(Vec<TypedStmt>),
    /// `x;`
    ExprStmt(TypedExpr),
    /// `;`
    EmptyStmt,
    /// `continue;`
    ContinueStmt,
    /// `break;`
    BreakStmt,
    /// `return;` or `return x;`
    ReturnStmt(Option<TypedExpr>),
    /// A let declaration
    DeclarationList(Vec<LetDeclaration>),
}

/// A struct or function declaration at the top level of a file
#[derive(Debug, Clone, PartialEq)]
pub enum TypedDeclaration {
    /// A declaration of a function
    FunctionDeclaration {
        /// The name of the function.
        name: String,
        /// The parameters of the function.
        parameters: Vec<ArgumentDeclaration>,
        /// The return type of the function. If set to [`None`], the function is
        /// void.
        return_type: Option<Type>,
        /// The body of the function. If set to [`None`], this is an extern
        /// declaration.
        body: Option<Vec<TypedStmt>>,
    },
    /// A named declaration for a `struct`.
    StructDeclaration {
        /// The name of the newtype.
        name: String,
        /// The key-value pairs of the struct. Ordered by declaration order.
        fields: IndexMap<String, super::ty::Type>,
    },
}

impl Display for TypedDeclaration {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::FunctionDeclaration {
                name,
                parameters,
                return_type: Some(r),
                body: Some(body),
            } => write!(
                f,
                "fn {name}({}) -> {r} {{\n{}\n}}",
                parameters
                    .iter()
                    .map(ToString::to_string)
                    .collect::<Vec<String>>()
                    .join(", "),
                body.iter()
                    .map(|stmt| stmt
                        .to_string()
                        .split('\n')
                        .map(|x| format!("    {x}"))
                        .collect::<Vec<_>>()
                        .join("\n"))
                    .collect::<Vec<String>>()
                    .join("\n")
            ),
            Self::FunctionDeclaration {
                name,
                parameters,
                return_type: Some(r),
                body: None,
            } => write!(
                f,
                "fn {name}({}) -> {r};",
                parameters
                    .iter()
                    .map(ToString::to_string)
                    .collect::<Vec<String>>()
                    .join(", ")
            ),
            Self::FunctionDeclaration {
                name,
                parameters,
                return_type: None,
                body: Some(body),
            } => write!(
                f,
                "fn {name}({}) {{\n{}\n}}",
                parameters
                    .iter()
                    .map(ToString::to_string)
                    .collect::<Vec<String>>()
                    .join(", "),
                body.iter()
                    .map(|stmt| stmt
                        .to_string()
                        .split('\n')
                        .map(|x| format!("    {x}"))
                        .collect::<Vec<_>>()
                        .join("\n"))
                    .collect::<Vec<String>>()
                    .join("\n")
            ),
            Self::FunctionDeclaration {
                name,
                parameters,
                return_type: None,
                body: None,
            } => write!(
                f,
                "fn {name}({});",
                parameters
                    .iter()
                    .map(ToString::to_string)
                    .collect::<Vec<String>>()
                    .join(", ")
            ),
            Self::StructDeclaration { name, fields } => write!(
                f,
                "struct {name} {{\n{}\n}}",
                fields
                    .iter()
                    .map(|(name, ty)| format!("    {name}: {ty}"))
                    .collect::<Vec<_>>()
                    .join(",\n")
            ),
        }
    }
}

impl Display for TypedStmt {
    #[allow(clippy::too_many_lines)]
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::IfStmt(e, s1, Some(s2)) => write!(
                f,
                "if ({e}) {{\n{}\n}} else {{\n{}\n}}",
                s1.iter()
                    .map(|stmt| stmt
                        .to_string()
                        .split('\n')
                        .map(|x| format!("    {x}"))
                        .collect::<Vec<_>>()
                        .join("\n"))
                    .collect::<Vec<_>>()
                    .join("\n"),
                s2.iter()
                    .map(|stmt| stmt
                        .to_string()
                        .split('\n')
                        .map(|x| format!("    {x}"))
                        .collect::<Vec<_>>()
                        .join("\n"))
                    .collect::<Vec<_>>()
                    .join("\n")
            ),
            Self::IfStmt(e, s1, None) => write!(
                f,
                "if ({e}) {{\n{}\n}}",
                s1.iter()
                    .map(|stmt| stmt
                        .to_string()
                        .split('\n')
                        .map(|x| format!("    {x}"))
                        .collect::<Vec<_>>()
                        .join("\n"))
                    .collect::<Vec<_>>()
                    .join("\n")
            ),
            Self::WhileStmt(e, s) => write!(
                f,
                "while ({e}) {{\n{}\n}}",
                s.iter()
                    .map(|stmt| stmt
                        .to_string()
                        .split('\n')
                        .map(|x| format!("    {x}"))
                        .collect::<Vec<_>>()
                        .join("\n"))
                    .collect::<Vec<_>>()
                    .join("\n")
            ),
            Self::ForStmt {
                init,
                cond,
                post,
                body,
            } => write!(
                f,
                "for ({} {}; {}) {{\n{}\n}}",
                init.clone().map_or(String::new(), |x| format!(
                    "let {};",
                    x.iter()
                        .map(std::string::ToString::to_string)
                        .collect::<Vec<_>>()
                        .join(", ")
                )),
                cond.clone().map_or(String::new(), |x| x.to_string()),
                post.clone().map_or(String::new(), |x| x.to_string()),
                body.iter()
                    .map(|stmt| stmt
                        .to_string()
                        .split('\n')
                        .map(|x| format!("    {x}"))
                        .collect::<Vec<_>>()
                        .join("\n"))
                    .collect::<Vec<_>>()
                    .join("\n")
            ),
            Self::BlockStmt(s) => write!(
                f,
                "{{\n{}\n}}",
                s.iter()
                    .map(|stmt| stmt
                        .to_string()
                        .split('\n')
                        .map(|x| format!("    {x}"))
                        .collect::<Vec<_>>()
                        .join("\n"))
                    .collect::<Vec<_>>()
                    .join("\n")
            ),
            Self::ExprStmt(e) => write!(f, "{e};"),
            Self::EmptyStmt => write!(f, ";"),
            Self::ContinueStmt => write!(f, "continue;"),
            Self::BreakStmt => write!(f, "break;"),
            Self::ReturnStmt(Some(e)) => write!(f, "return {e};",),
            Self::ReturnStmt(None) => write!(f, "return;"),
            Self::DeclarationList(d) => write!(
                f,
                "let {};",
                d.iter()
                    .map(std::string::ToString::to_string)
                    .collect::<Vec<_>>()
                    .join(", ")
            ),
        }
    }
}

/// A special form of [`LetDeclaration`] used for function parameters.
#[derive(PartialEq, Debug, Clone)]
pub struct ArgumentDeclaration {
    /// The name of the parameter.
    pub name: String,
    /// The type of the parameter.
    pub ty: Type,
}
impl Display for ArgumentDeclaration {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}: {}", self.name, self.ty)
    }
}
