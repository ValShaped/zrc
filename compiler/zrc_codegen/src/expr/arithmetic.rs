//! code generation for arithmetic expressions

use inkwell::{
    builder::BuilderError,
    values::{BasicValue, BasicValueEnum, IntValue},
};
use zrc_typeck::tast::{
    expr::{Arithmetic, BinaryBitwise, TypedExpr},
    ty::Type,
};

use crate::{
    bb::{BasicBlockAnd, BasicBlockExt},
    ctx::BlockCtx,
    expr::{CgExprArgs, cg_expr},
    ty::llvm_basic_type,
    unpack,
};

/// Build the required instruction for a [`BinaryBitwise`] operation
pub fn build_binary_bitwise<'ctx>(
    cg: BlockCtx<'ctx, '_, '_>,
    op: BinaryBitwise,
    lhs: IntValue<'ctx>,
    rhs: IntValue<'ctx>,
    result_is_signed: bool,
) -> Result<IntValue<'ctx>, BuilderError> {
    match op {
        BinaryBitwise::And => cg.builder.build_and(lhs, rhs, "and"),
        BinaryBitwise::Or => cg.builder.build_or(lhs, rhs, "or"),
        BinaryBitwise::Xor => cg.builder.build_xor(lhs, rhs, "xor"),
        BinaryBitwise::Shl => cg.builder.build_left_shift(lhs, rhs, "shl"),
        BinaryBitwise::Shr => cg
            .builder
            .build_right_shift(lhs, rhs, result_is_signed, "shr"),
    }
}
/// Build the required instruction for an [`Arithmetic`] operation
pub fn build_arithmetic<'ctx>(
    cg: BlockCtx<'ctx, '_, '_>,
    op: Arithmetic,
    lhs: IntValue<'ctx>,
    rhs: IntValue<'ctx>,
    result_is_signed: bool,
) -> Result<IntValue<'ctx>, BuilderError> {
    match op {
        Arithmetic::Addition => cg.builder.build_int_add(lhs, rhs, "add"),
        Arithmetic::Subtraction => cg.builder.build_int_sub(lhs, rhs, "sub"),
        Arithmetic::Multiplication => cg.builder.build_int_mul(lhs, rhs, "mul"),

        Arithmetic::Division if result_is_signed => {
            cg.builder.build_int_signed_div(lhs, rhs, "div")
        }
        Arithmetic::Division => cg.builder.build_int_unsigned_div(lhs, rhs, "div"),

        Arithmetic::Modulo if result_is_signed => cg.builder.build_int_signed_rem(lhs, rhs, "rem"),
        Arithmetic::Modulo => cg.builder.build_int_unsigned_rem(lhs, rhs, "rem"),
    }
}

/// Code generate a binary bitwise operation
pub fn cg_binary_bitwise<'ctx, 'input>(
    CgExprArgs {
        cg,
        mut bb,
        inferred_type,
        ..
    }: CgExprArgs<'ctx, 'input, '_>,
    op: BinaryBitwise,
    lhs: Box<TypedExpr<'input>>,
    rhs: Box<TypedExpr<'input>>,
) -> BasicBlockAnd<'ctx, BasicValueEnum<'ctx>> {
    let lhs = unpack!(bb = cg_expr(cg, bb, *lhs));
    let rhs = unpack!(bb = cg_expr(cg, bb, *rhs));

    let reg = build_binary_bitwise(
        cg,
        op,
        lhs.into_int_value(),
        rhs.into_int_value(),
        inferred_type.is_signed_integer(),
    )
    .expect("binary bitwise operation should have compiled successfully");

    bb.and(reg.as_basic_value_enum())
}

/// Code generate an arithmetic operation
pub fn cg_arithmetic<'ctx, 'input>(
    CgExprArgs {
        cg,
        mut bb,
        inferred_type,
        ..
    }: CgExprArgs<'ctx, 'input, '_>,
    op: Arithmetic,
    lhs: Box<TypedExpr<'input>>,
    rhs: Box<TypedExpr<'input>>,
) -> BasicBlockAnd<'ctx, BasicValueEnum<'ctx>> {
    let lhs_ty = lhs.inferred_type.clone();
    let lhs = unpack!(bb = cg_expr(cg, bb, *lhs));
    let rhs = unpack!(bb = cg_expr(cg, bb, *rhs));

    if let Type::Ptr(pointee) = lhs_ty {
        // Most languages make incrementing a pointer increase the address by the size
        // of the pointee type, hence our use of `gep`.
        // REVIEW: Is this the approach we want to take?
        #[allow(clippy::wildcard_enum_match_arm)]
        let reg = match op {
            // SAFETY: This can segfault if indices are used incorrectly
            // This is only used for pointer arithmetic, so the indices should be correct
            Arithmetic::Addition => unsafe {
                cg.builder.build_gep(
                    llvm_basic_type(&cg, &pointee).0,
                    lhs.into_pointer_value(),
                    &[rhs.into_int_value()],
                    "ptr_add",
                )
            },
            Arithmetic::Subtraction => {
                let rhs = cg
                    .builder
                    .build_int_neg(rhs.into_int_value(), "neg")
                    .expect("negation should have compiled successfully")
                    .as_basic_value_enum();

                // SAFETY: This can segfault if indices are used incorrectly
                // This is only used for pointer arithmetic, so the indices should be
                // correct
                unsafe {
                    cg.builder.build_gep(
                        llvm_basic_type(&cg, &pointee).0,
                        lhs.into_pointer_value(),
                        &[rhs.into_int_value()],
                        "ptr_sub",
                    )
                }
            }
            _ => panic!("invalid pointer arithmetic operation"),
        }
        .expect("pointer arithmetic should have compiled successfully");

        bb.and(reg.as_basic_value_enum())
    } else {
        let reg = build_arithmetic(
            cg,
            op,
            lhs.into_int_value(),
            rhs.into_int_value(),
            inferred_type.is_signed_integer(),
        )
        .expect("arithmetic operation should have compiled successfully");

        bb.and(reg.as_basic_value_enum())
    }
}

/// Code generate a unary bitwise NOT operation
pub fn cg_unary_bitwise_not<'ctx, 'input>(
    CgExprArgs { cg, mut bb, .. }: CgExprArgs<'ctx, 'input, '_>,
    x: Box<TypedExpr<'input>>,
) -> BasicBlockAnd<'ctx, BasicValueEnum<'ctx>> {
    let value = unpack!(bb = cg_expr(cg, bb, *x));

    let reg = cg
        .builder
        .build_not(value.into_int_value(), "not")
        .expect("not should have compiled successfully");

    bb.and(reg.as_basic_value_enum())
}

/// Code generate a unary minus operation
pub fn cg_unary_minus<'ctx, 'input>(
    CgExprArgs { cg, mut bb, .. }: CgExprArgs<'ctx, 'input, '_>,
    x: Box<TypedExpr<'input>>,
) -> BasicBlockAnd<'ctx, BasicValueEnum<'ctx>> {
    let value = unpack!(bb = cg_expr(cg, bb, *x));

    let reg = cg
        .builder
        .build_int_neg(value.into_int_value(), "neg")
        .expect("negation should have compiled successfully");

    bb.and(reg.as_basic_value_enum())
}

#[cfg(test)]
mod tests {
    // Please read the "Common patterns in tests" section of crate::test_utils for
    // more information on how code generator tests are structured.

    use indoc::indoc;

    use crate::cg_snapshot_test;

    #[test]
    fn pointer_arithmetic_generates_proper_gep() {
        cg_snapshot_test!(indoc! {r#"
                fn test() {
                    let x: *i32;

                    // TEST: should create a GEP that is the same as "x[4]"
                    let y = x + 4 as usize;
                    // TEST: and the same, with -4:
                    let z = x - 4 as usize;
                }
            "#});
    }

    #[test]
    fn arithmetic_operators_generate() {
        cg_snapshot_test!(indoc! {"
                fn get_int() -> i32;
                fn get_uint() -> u32;

                fn test() {
                    let sx = get_int();
                    let sy = get_int();
                    let ux = get_uint();
                    let uy = get_uint();

                    // TEST: should create an `add i32` instruction
                    let s_add = sx + sy;
                    let u_add = ux + uy;

                    // TEST: should create a `sub i32` instruction
                    let s_sub = sx - sy;
                    let u_sub = ux - uy;

                    // TEST: should create a `mul i32` instruction
                    let s_mul = sx * sy;
                    let u_mul = ux * uy;

                    // TEST: should create a `sdiv i32` instruction
                    let s_div = sx / sy;
                    // TEST: should create a `udiv i32` instruction
                    let u_div = ux / uy;

                    // TEST: should create a `srem i32` instruction
                    let s_rem = sx % sy;
                    // TEST: should create a `urem i32` instruction
                    let u_rem = ux % uy;
                }
            "});
    }

    #[test]
    fn bitwise_operators_generate() {
        cg_snapshot_test!(indoc! {"
                fn get_int() -> i32;
                fn get_uint() -> u32;

                fn test() {
                    let x = get_int();
                    let y = get_int();
                    let u = get_uint();

                    // TEST: should create a `not i32` instruction
                    let not = ~x;

                    // TEST: should create a `and i32` instruction
                    let and = x & y;

                    // TEST: should create a `or i32` instruction
                    let or = x | y;

                    // TEST: should create a `xor i32` instruction
                    let xor = x ^ y;

                    // TEST: should create a `shl i32` instruction
                    let shl = x << u;

                    // TEST: should create a `lshr i32` instruction
                    let lshr = u >> u;

                    // TEST: should create a `ashr i32` instruction (as the lhs is signed)
                    let ashr = x >> u;
                }
            "});
    }
}
