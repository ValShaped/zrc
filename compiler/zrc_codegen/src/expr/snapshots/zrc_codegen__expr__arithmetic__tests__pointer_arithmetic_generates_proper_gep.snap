---
source: compiler/zrc_codegen/src/expr/arithmetic.rs
description: "fn test() {\n    let x: *i32;\n\n    // TEST: should create a GEP that is the same as \"x[4]\"\n    let y = x + 4 as usize;\n    // TEST: and the same, with -4:\n    let z = x - 4 as usize;\n}\n"
expression: resulting_ir
---
; ModuleID = 'test.zr'
source_filename = "test.zr"

define {} @test() !dbg !3 {
entry:
  %let_z = alloca ptr, align 8
  %let_y = alloca ptr, align 8
  %let_x = alloca ptr, align 8
  %load = load ptr, ptr %let_x, align 8, !dbg !8
  %ptr_add = getelementptr i32, ptr %load, i64 4, !dbg !11
  store ptr %ptr_add, ptr %let_y, align 8, !dbg !12
  %load1 = load ptr, ptr %let_x, align 8, !dbg !13
  %ptr_sub = getelementptr i32, ptr %load1, i64 -4, !dbg !14
  store ptr %ptr_sub, ptr %let_z, align 8, !dbg !15
  ret {} zeroinitializer, !dbg !16
}

!llvm.module.flags = !{!0}
!llvm.dbg.cu = !{!1}

!0 = !{i32 2, !"Debug Info Version", i32 3}
!1 = distinct !DICompileUnit(language: DW_LANG_C, file: !2, producer: "zrc test runner", isOptimized: false, flags: "zrc --fake-args", runtimeVersion: 0, emissionKind: FullDebug, splitDebugInlining: false)
!2 = !DIFile(filename: "test.zr", directory: "/fake/path")
!3 = distinct !DISubprogram(name: "test", linkageName: "test", scope: null, file: !2, line: 1, type: !4, scopeLine: 1, spFlags: DISPFlagDefinition, unit: !1)
!4 = !DISubroutineType(types: !5)
!5 = !{!6}
!6 = !DICompositeType(tag: DW_TAG_structure_type, name: "struct {}", scope: !2, file: !2, elements: !7)
!7 = !{}
!8 = !DILocation(line: 5, column: 13, scope: !9)
!9 = distinct !DILexicalBlock(scope: !10, file: !2, line: 1, column: 11)
!10 = distinct !DILexicalBlock(scope: !3, file: !2, line: 1, column: 11)
!11 = !DILocation(line: 5, column: 17, scope: !9)
!12 = !DILocation(line: 5, column: 9, scope: !9)
!13 = !DILocation(line: 7, column: 13, scope: !9)
!14 = !DILocation(line: 7, column: 17, scope: !9)
!15 = !DILocation(line: 7, column: 9, scope: !9)
!16 = !DILocation(line: 8, column: 1, scope: !9)
